Remotion/Data/DomainObjects/Persistence/Rdbms/Parameters/ObjectIDDataParameterDefinitionFactory.cs

```diff
@@ -0,0 +1,82 @@
+//
+using Remotion.Data.DomainObjects.Persistence.Configuration;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model.Building;
+using Remotion.Data.DomainObjects.Queries;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+
+/// <summary>
+/// Can create <see cref="IDataParameterDefinition"/> instances that can handle <see cref="ObjectID"/> parameter values.
+/// </summary>
+public class ObjectIDDataParameterDefinitionFactory : IDataParameterDefinitionFactory
+{
+  public IStorageSettings StorageSettings { get; }
+  public StorageProviderDefinition StorageProviderDefinition { get; }
+  public IStorageTypeInformationProvider StorageTypeInformationProvider { get; }
+  public IDataParameterDefinitionFactory NextDataParameterDefinitionFactory { get; }
+
+  public ObjectIDDataParameterDefinitionFactory (StorageProviderDefinition storageProviderDefinition,
+      IStorageTypeInformationProvider storageTypeInformationProvider,
+      IStorageSettings storageSettings,
+      IDataParameterDefinitionFactory nextDataParameterDefinitionFactory)
+  {
+    ArgumentUtility.CheckNotNull(nameof(storageProviderDefinition), storageProviderDefinition);
+    ArgumentUtility.CheckNotNull(nameof(storageTypeInformationProvider), storageTypeInformationProvider);
+    ArgumentUtility.CheckNotNull(nameof(storageSettings), storageSettings);
+    NextDataParameterDefinitionFactory = nextDataParameterDefinitionFactory;
+
+    StorageSettings = storageSettings;
+    StorageProviderDefinition = storageProviderDefinition;
+    StorageTypeInformationProvider = storageTypeInformationProvider;
+  }
+
+  public IDataParameterDefinition CreateDataParameterDefinition (ObjectID objectID)
+  {
+    ArgumentUtility.CheckNotNull(nameof(objectID), objectID);
+
+    var relatedStorageProviderDefinition = StorageSettings.GetStorageProviderDefinition(objectID.ClassDefinition);
+
+    if (StorageProviderDefinition != relatedStorageProviderDefinition)
+      return CreateCrossProviderRelationStoragePropertyDefinition();
+    else
+     return CreateSameProviderRelationStoragePropertyDefinition();
+  }
+
+  private IDataParameterDefinition CreateSameProviderRelationStoragePropertyDefinition ()
+  {
+    var storageTypeInformation = StorageTypeInformationProvider.GetStorageTypeForID(true);
+    return new ObjectIDDataParameterDefinition(storageTypeInformation);
+  }
+
+  private IDataParameterDefinition CreateCrossProviderRelationStoragePropertyDefinition ()
+  {
+    var storageTypeInfo = StorageTypeInformationProvider.GetStorageTypeForSerializedObjectID(true);
+    return new SerializedObjectIDDataParameterDefinition(storageTypeInfo);
+  }
+
+  public IDataParameterDefinition CreateDataParameterDefinition (QueryParameter queryParameter)
+  {
+    var objectID = queryParameter.Value as ObjectID;
+    if (objectID != null)
+    {
+      return CreateDataParameterDefinition(objectID);
+    }
```

Comment:
```suggestion
    if (queryParameter.Value is ObjectID objectID)
      return CreateDataParameterDefinition(objectID);
```

In case you want to use the new, more compact syntax.